---
title: "htapp_methods_NSCLC_inferCNV.Rmd"
author: "Gabriela Smith-Rosario"
date: "4/26/2019"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---

Obtain copy nummber variation of NSCLC cancer data.

```{r}
rm(list = ls())

# Load libraries
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())
# library(refGenome)
library(plyr)
library(dplyr)
library(Seurat)
library(infercnv)
library(rlist)
library(slam)
library(tibble)
library(ggpubr)
library(Matrix)
library(gplots)
library(GSA)
library(stringr)

# Set directories 
proj.path <- "/path/to/project/directory"
user.path <- "/path/to/user/files"

# Set date year_month_day to use in figure directory name and in names for saving Rda objects.
date <- "ENTER_DATE"  
name.project <- "HTAPP_NSCLC_example_inferCNV"

# Load Seurat object that was converted from cumulus.
load(paste0(proj.path, "/src/Rdata/ENTER_DATE_HTAPP_NSCLC_example_seurat_annotate.Rda"))

# Read in cumulus results 
sampleid.cumulus <- c("NSCLC14_C4", "NSCLC14_PDEC", "NSCLC14_LE", "NSCLC14_combined")

# Create output directories.
figures.dir <- paste0(proj.path, "/results/NSCLC/", date, "_inferCNV/")
figures.path <- sapply(sampleid.cumulus, function(i) paste0(figures.dir, i), USE.NAMES = F) 

# File containing raw counts matrix for both cells to infer CNV for and cells to use as the reference and
# file containing annotation of each cell as malignant or non-malignant.
file.counts <- paste0(figures.path, "/", sampleid.cumulus, "_counts.txt")
file.annotation <- paste0(figures.path, "/", sampleid.cumulus, "_annotation.txt")

# File generated by gtf_to_position_file.py containing gene chromosome, start, and end positions.
file.genomic.position <- paste0(proj.path, "/doc/GRCh38_genomic_positions.txt")
file.shuffle.genomic.position <- paste0(proj.path, "/doc/GRCh38_shuffle_genomic_positions.txt")

# Save inferCNV outputs
Rda.inferCNV.normaltumor.path <- paste0(proj.path, "/src/Rdata/", date , "_10x_", name.project, ".Rda") 

out.dir <- c(figures.dir, figures.path, paste0(proj.path, "/src/Rdata"), paste0(figures.dir, "tSNE_scores/"))
sapply(out.dir, function(i) {if (!dir.exists(i)) {dir.create(i, recursive = T)} })

# Load my functions.
source(paste0(user.path, "/code/R/plotutils.R"))
source(paste0(user.path, "/code/R/seuratutils.R"))
source(paste0(user.path, "/code/R/color.R"))
```

## Fix duplicated gene names in genomic coordinate position file and make a shuffled position file
I found that some of the gene names are not unique in the initial genomic positions file. There are not many duplicates, for instance 14 genes in GRCh38, but any duplicates will cause problems in `inferCNV` so I also wrote `R` code to take these files and make the gene names unique. One gene name is unchanged and the duplicate is altered with an underscore and number. Because there are so few duplicates, this should not significantly affect the analysis. There are 20,419 gene name duplicates in hg19, which I think may be mostly in non-coding RNA genes.

Make any duplicated gene names in genomic position file unique. I set evaluation of this code to False because it only needs to be run once, as I overwrite the original file with the new file with unique gene names.
```{r positions, eval = F}
# Read in positions file
f <- read.table(file.genomic.position, sep = '\t', stringsAsFactors = F)
autosome <- f[which(f$V2 %in% c(1:22),),]
autosome$V2 <- as.integer(autosome$V2)
autosome <- setorder(autosome, cols = V2)
sex.chromosome <- f[which(f$V2 %in% c("X","Y")),]
other <- f[which(!f$V2 %in% c("X","Y", c(1:22))),]
other <- setorder(other, cols = V2)

position <- rbind(autosome,sex.chromosome, other)

# Make duplicated gene names unique and write new positions file.
position$V1 <- make.unique(position$V1)  # make any duplicated gene names unique
write.table(position, file.genomic.position, sep = "\t", row.names = F, col.names = F, quote = F)

# Fix order of chromosomes to 1,2,3,...20,21,X,Y,M.

# Shuffle gene names so that they have different chromosome numbers and start/end positions.
position$V1 <- sample(position$V1)
write.table(position, file.shuffle.genomic.position, sep = "\t", row.names = F, col.names = F, quote = F)
```

## Prepare count matrix file and annotation file (clustering from cumulus) on a subsample of the cells 
inferCNV takes in the raw counts matrices from cells to infer CNV for, and from the reference cells. It needs these count for all these cells in a single counts matrix. It also needs the labels of the cells in that count matrix, ie are these cells to infer CNV for or are these cells to use as the reference?

```{r make_CNV_input, eval = F}
sample.name <- as.vector(sampleid.cumulus)  #The number of samples

sapply(gcdata, function(i) ncol(i@data)) # Look at the number of cells/nuclei per sample

## Select which cell types to use as a reference
reference.annotation <- c("B cell", "NK cell", "T cell", "Macrophage", "Mast cell", "Endothelial cell")

gcdata.tumor <- list()
gcdata.normal <- list()

for (i in seq(sampleid.cumulus)){
  gc <- gcdata[[i]]
  gc@meta.data$orig.names <- gc@cell.names  # save original cell names in meta.data
  
  # Label cells coming from reference and cells coming from sample.
  cells.non.malignant  <- rownames(gc@meta.data[which(gc@meta.data$annotate %in% reference.annotation),])
  cells.malignant <- rownames(gc@meta.data[which(!(gc@meta.data$annotate %in% reference.annotation)),])
  gc@meta.data[cells.malignant, "inferCNV.annotate"] <- "malignant"
  gc@meta.data[cells.non.malignant, "inferCNV.annotate"] <- "non-malignant"
  
  if (length(cells.non.malignant) > 0 && length(cells.malignant)){
    
    
    # Save count matrix. Save @raw.data, which is raw UMI data and not log transformed
    gcdata.non.malignant <- SubsetData(gc, cells.use = cells.non.malignant)
    gcdata.non.malignant <- RenameCells(gcdata.non.malignant, add.cell.id = "non-malignant")  
    gcdata.non.malignant@meta.data$orig.names <- gcdata.non.malignant@cell.names  # save original cell names in meta.data
    if (length(gcdata.non.malignant@cell.names) < 2000){
      cells.use <- gcdata.non.malignant@cell.names
    }else{
      cells.use <- sample(gcdata.non.malignant@cell.names, 2000) # subsample to save time in testing code
      gcdata.non.malignant <- SubsetData(gcdata.non.malignant, cells.use = cells.use)
    }
    gcdata.non.malignant@raw.data <- exp(gcdata.non.malignant@data) - 1  # infercnv expects UMI counts or TPM, and it normalizes these
    Matrix::colSums(gcdata.non.malignant@raw.data) %>% range
    gcdata.normal[[i]] <- gcdata.non.malignant
    
    gcdata.malignant <- SubsetData(gc, cells.use = cells.malignant)
    gcdata.malignant <- RenameCells(gcdata.malignant, add.cell.id = "malignant")  
    gcdata.malignant@meta.data$orig.names <- gcdata.malignant@cell.names  # save original cell names in meta.data
    if (length(gcdata.malignant@cell.names) < 2000){
      cells.use <- gcdata.malignant@cell.names
    }else{
      cells.use <- sample(gcdata.malignant@cell.names, 2000) # subsample to save time in testing code
      gcdata.malignant <- SubsetData(gcdata.malignant, cells.use = cells.use)
    }
    gcdata.malignant@raw.data <- exp(gcdata.malignant@data) - 1  # infercnv expects UMI counts or TPM, and it normalizes these
    Matrix::colSums(gcdata.malignant@raw.data) %>% range
    gcdata.tumor[[i]] <- gcdata.malignant
    
    # Save count matrix. Save @raw.data, which is raw UMI data and not log transformed.
    genes.intersect <- intersect(rownames(gcdata.malignant@raw.data), rownames(gcdata.non.malignant@raw.data)) # genes   shared between reference and sample
    tpm.non.malignant <- as.matrix(gcdata.non.malignant@raw.data[genes.intersect, gcdata.non.malignant@cell.names])
    tpm.malignant <- as.matrix(gcdata.malignant@raw.data[genes.intersect, gcdata.malignant@cell.names])
    all.equal(rownames(tpm.non.malignant), rownames(tpm.malignant)) #Verify that the number of genes and order is the same
    tpm <- cbind(tpm.non.malignant, tpm.malignant)
    
    # Write count matrix to file.
    write.table(round(tpm, digits = 3), file.counts[i], sep = "\t", row.names = T, col.names = T, quote = F)
    
    # Label cells coming from reference and cells coming from sample.
    labels.non.malignant <- gcdata.non.malignant@meta.data[, c("orig.names", "annotate")]
    labels.malignant <- gcdata.malignant@meta.data[, c("orig.names", "annotate")]
    labels <- rbind(labels.non.malignant, labels.malignant)
    all.equal(labels$orig.names, colnames(tpm))
    
    # Write annotation of cells (malignant or non-malignant) to file.
    write.table(labels, file.annotation[i], quote = F, row.names = F, col.names = F, sep = "\t")
  } 
}
```

## Create the InferCNV Object and run InferCNV on the subsampled data
Read in the raw counts matrix, annotation file, and gene position file along chromosomes.
```{r create_inferCNV, eval = T}
for (i in seq(sampleid.cumulus)){
  if (length(list.files(figures.path[i])) > 1){
    
    ann.use <- read.csv(paste0(file.annotation[i]), sep = "\t")
    
    ref.use <- c("B cell", "NK cell", "T cell", "Macrophage", "Mast cell", "Endothelial cell")[which(c("B cell", "NK cell", "T cell", "Macrophage", "Mast cell", "Endothelial cell") %in% unique(ann.use[,2]))]
    
    
    infercnv_obj = CreateInfercnvObject(
      raw_counts_matrix = file.counts[i],
      annotations_file = file.annotation[i],
      delim="\t",
      gene_order_file = file.genomic.position,
      ref_group_names = ref.use)
    
    out_dir <- figures.path[i]
    # Perform infercnv operations to reveal cnv signal.
    infercnv_obj = infercnv::run(infercnv_obj,
                                 cutoff=0.1, 
                                 out_dir=out_dir, 
                                 cluster_by_groups=T, 
                                 plot_steps=F,
                                 denoise=T,
                                 resume_mode = T,
                                 HMM=T,
                                 HMM_type = "i6",
                                 num_threads = 8
    )
  }
}
```

## Identify malignant cells from the data based on the inferCNV observations output files
```{r identify_malignant_cells, eval = T}

## Save CNV values as metadata

gcdata.cnv <- list()
gcdata.cnv.names <- c()

# Get magnitude of CNV summed for each cell. First I take the log, then absolute value, then sum across genomic positions.
for (i in seq(gcdata)){
  if (length(list.files(figures.path[i])) > 1){
    
    obs <- read.table(paste0(figures.path[i], "/infercnv.observations.txt"), stringsAsFactors = F)
    if("TRUE" %in% (grepl(rownames(obs), pattern = "(malignant)"))){
      obs <- t(obs)
    }
    colnames(obs) <- gsub(colnames(obs), pattern = "[.]", replacement = "-")
    cnv.malignant <- apply(obs, 2, function(i) sum(abs(log2(i))))
    intersect(gcdata[[i]]@cell.names, gsub(colnames(obs), pattern = "(malignant_)", replacement = "")) %>% length
    cnv.malignant <- as.matrix(cnv.malignant)
    rownames(cnv.malignant) <- gsub(rownames(cnv.malignant), pattern = "(malignant_)", replacement = "")
    
    ref <- read.table(paste0(figures.path[i], "/infercnv.references.txt"), stringsAsFactors = F)
    if("TRUE" %in% (grepl(rownames(ref), pattern = "(malignant)"))){
      ref <- t(ref)
    }
    colnames(ref) <- gsub(colnames(ref), pattern = "[.]", replacement = "-")
    cnv.nonmalignant <- apply(ref, 2, function(i) sum(abs(log2(i))))
    intersect(gcdata[[i]]@cell.names, gsub(colnames(ref), pattern = "(non-malignant_)", replacement = "")) %>% length
    cnv.nonmalignant <- as.matrix(cnv.nonmalignant)
    rownames(cnv.nonmalignant) <- gsub(rownames(cnv.nonmalignant), pattern = "(non-malignant_)", replacement = "")
    
    cnv.all <- rbind(cnv.malignant, cnv.nonmalignant)
    
    gcdata.sub <- SubsetData(gcdata[[i]], cells.use = rownames(cnv.all))
    gcdata.cnv.names[i] <- names(gcdata)[i]
    
    gcdata.sub@meta.data$cnv <- cnv.all[gcdata.sub@cell.names,]  # add CNV magnitude to Seurat object metadata
    MultipleFeaturePlot(gcdata.sub, features = "cnv", feature.type = "meta", pt.size = 2)
    ggsave(paste0(figures.path[i], "/tSNE_CNV_magnitude.png"), width = 8, height = 8)
    
    gcdata.cnv[[i]] <- gcdata.sub
  }
}

names(gcdata.cnv) <- unlist(gcdata.cnv.names)

save(gcdata.cnv, file = Rda.inferCNV.normaltumor.path)
```