---
title: "HTAPP_Methods_NSCLC"
author: "Caroline Porter"
date: "4/18/2019"
output: html_document
---

# Load libraries, set paths, etc. 
```{r}

# Load libraries
rm(list = ls())
library(reticulate)  
# load python environment 
use_python("/users/USER_NAME/software/miniconda3/envs/ENV_NAME/bin/python", required = T)
use_condaenv("ENV_NAME", conda = "/users/USER_NAME/software/miniconda3/bin/conda")
py_config()  # check that the expected python is loaded for scrublet


library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())
library(Seurat)
library(dplyr)
library(Matrix)
library(gplots)
library(GSA)
library(stringr)
library(reticulate)
library(DropletUtils)

# Set directories 
proj.path <- "/path/to/project/directory"
user.path <- "/path/to/user/files"

# Set date year_month_day to use in figure directory name and in names for saving Rda objects.
date <- "ENTER_DATE"  
name.project <- "HTAPP_NSCLC_example"

### Read in raw counts data (pre-cumulus/Cumulus) 
# List of the sample IDs for which count matrices were generated 
sampleid.counts <- c("NSCLC14_C4", "NSCLC14_PDEC", "NSCLC14_LE") # NEED TO UPDATE THESE 
prefix.counts <- c("/alignreads/NSCLC14/")  
expression.path <- sapply(seq(sampleid.counts), function(i) paste0(proj.path, prefix.counts, sampleid.counts[i], "/"))
expression.files <- sapply(expression.path, function(i) Sys.glob(paste0(i, "raw*h5")))
names(expression.path) <- sampleid.counts
names(expression.files) <- sampleid.counts

### Read in cumulus/Cumulus results 
sampleid.cumulus <- c("NSCLC14_C4", "NSCLC14_PDEC", "NSCLC14_LE", "NSCLC14_combined")
prefix.cumulus <- c("/results/NSCLC14/")
h5ad.file <- sapply(sampleid.cumulus, function(i) Sys.glob(paste0(proj.path, prefix.cumulus, i, '/', "*seurat.h5ad")))
names(h5ad.file) <- sampleid.cumulus

# List of 10x metrics_summary.csv and molecule_info.h5 files containing quality control metrics.
qc.files <- sapply(expression.path, function(i) Sys.glob(paste0(i, "/metrics*csv")))
h5.files <- sapply(expression.path, function(i) Sys.glob(paste0(i, "/molecule*h5")))
names(qc.files) <- sampleid.counts
names(h5.files) <- sampleid.counts 

# Save paths
Rda.cumulus.path <- paste0(proj.path, "/src/Rdata/", date, "_10x_", name.project, "_convert_scrtools_to_seurat_cumulus.Rda")
Rda.emptydrops.path <- paste0(proj.path, "/src/Rdata/", date, "_10x_", name.project, "_emptydrops.Rda")  # run empty drops
Rda.cumulus.scrublet.path <- paste0(proj.path, "/src/Rdata/", date, "_10x_", name.project, "_cumulus_scrublet.Rda")  # run both empty drops and remove doublets
Rda.cumulus.scrublet.emptydrops.path <- paste0(proj.path, "/src/Rdata/", date, "_10x_", name.project, "_cumulus_scrublet_emptydrops.Rda")  # save seurat object with emptydrops and scrublet meta data 
Rda.emptydrops.stats <- paste0(proj.path, "/src/Rdata/", date, "_10x_", name.project, "_emptydrops_stats.Rda")  # number of cells and genes per sample before and after EmptyDrops run

# Create output directories.
figures.dir <- paste0(proj.path, "/results/NSCLC/", date, "_10x_cumulus/")
figures.final <- paste0(proj.path, "/results/NSCLC/Final_Figures/")
figures.path <- sapply(sampleid.cumulus, function(i) paste0(figures.dir, i, "_"), USE.NAMES = F) 
figures.path.counts <- sapply(sampleid.counts, function(i) paste0(figures.dir, i, "_"), USE.NAMES = F) 
sparse.dir <- sapply(sampleid.cumulus, function(i) paste0(figures.dir, "matrix_emptydrops", "/", i, "/"), USE.NAMES = F) 
figures.pythondir <- figures.dir  # where to write figures in python chunk

out.dir <- c(figures.dir, figures.final, sparse.dir, paste0(proj.path, "/src/Rdata"))
sapply(out.dir, function(i) {if (!dir.exists(i)) {dir.create(i, recursive = T)} })

# Load my functions.
source(paste0(user.path, "/code/orr_plotutils_04182019.R"))
source(paste0(user.path, "/code/orr_seuratutils_04182019.R"))
source(paste0(user.path, "/code/orr_color_04182019.R"))

# knitr settings for code chunks.
knitr::opts_knit$set(root.dir = proj.path)  # set working directory for all chunks
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
```

## Load raw data and run empty drops on the raw counts data; save progress 
```{r init, eval = T}
# Load UMI data from each Cell Ranger gene expression matrix into a list. The list elements are named
# by their sample name. When the h5 file is read in, the GEM group is kept with the cell barcode (16nt-1).
# We remove the GEM group so as to match the cell barcode names in molecule_info.h5.
gcdata.data <- sapply(expression.files, function(dir) Read10X_h5(filename = dir))
gcdata.data <- sapply(seq(gcdata.data), function (i) {colnames(gcdata.data[[i]]) <- gsub("-[0-9]+", "", colnames(gcdata.data[[i]]), perl = T); gcdata.data[[i]]})  # remove GEM group from cell barcode

# Identify empty droplets using DropletUtils from the Marioni lab. JUST FLAG EMPTY DROPS
set.seed(100)  # EmptyDrops is written in R and we want reproducible runs
cells.keep.emptydrops <- sapply(seq(gcdata.data), function(i) RunEmptyDrops(gcdata.data[[i]], figures.path.counts[i], title = sampleid.counts[i]))  # cell barcodes that pass EmptyDrops filter in each 10x channel
names(gcdata.data) <- sampleid.counts  # add list names for each counts matrix
names(cells.keep.emptydrops) <- sampleid.counts

# Save current progress
save(gcdata.data, sampleid.counts, cells.keep.emptydrops, file = Rda.emptydrops.path)

```

## Read in h5ad files and convert them from anndata to Seurat objects; save progress 
```{r anndata_seurat, eval = F}
ad <- import("anndata", convert = FALSE)  # import python module anndata
sci <- import("scipy", convert = FALSE)  # import python module scipy
np <- import("numpy", convert = FALSE)  # import python module numpy

# Read in a list of anndata objects and convert them to Seurat objects.
gcdata.ad <- sapply(h5ad.file, function(h5ad) ad$read_h5ad(h5ad))
gcdata <- list()
for (i in seq(gcdata.ad)){
  if ("numpy.ndarray" %in% class(gcdata.ad[[i]]$X)){
    gcdata[[i]] <- Convert(gcdata.ad[[i]], to = "seurat")
  }else{
  source("https://raw.githubusercontent.com/klarman-cell-observatory/KCO/master/workflows/scCloud/h5ad2seurat.R")
  gcdata[[i]] <- Convert.anndata.base.AnnData(gcdata.ad[[i]], to = "seurat")
  }
}

names(gcdata) <- names(gcdata.ad)

# Assign gcdata@ident based on Louvain labels and add labels to meta data
gcdata <- sapply(gcdata, function(data) SetAllIdent(data, id = "louvain_labels"))
cumulus.ident <- sapply(gcdata, function(data) data@meta.data$louvain_labels)

# Save current progress
save(gcdata, file = Rda.cumulus.path)

# Check that @data is normalized, log-transformed data
# and that the mean expression for each gene in @scale.data is 0.
i <- length(gcdata)
Matrix::rowSums(gcdata[[i]]@scale.data) %>% range
Matrix::rowMeans(gcdata[[i]]@scale.data) %>% range
Matrix::colSums(exp(gcdata[[i]]@data)-1) %>% range


```

## Prepare count matrices for doublet identification
This chunk writes the count matrices to file so that they may be read in later by scrublet, which runs in python.
Scrublet will be run on the @raw.data slot from the cumulus output, after conversion to Seurat object 
```{r prepare_scrublet, eval = T}

load(Rda.cumulus.path)

# Write the counts matrix, containing counts after filtering the empty drops, in Market Matrix format.
for (i in seq(sampleid.cumulus)) {
  cells.keep <- gcdata[[sampleid.cumulus[[i]]]]@cell.names  # only use cells that passed filtering
  rawdata <- gcdata[[sampleid.cumulus[i]]]@raw.data[, cells.keep]
  write10xCounts(sparse.dir[i], rawdata, barcodes=colnames(rawdata), gene.id=rownames(rawdata), overwrite=T)
  # write10xCounts(sparse.dir[i], rawdata, barcodes=colnames(rawdata), gene.id=rownames(rawdata), overwrite=T, type="sparse")
}
# rawdata <- as.matrix(gcdata[[2]]@raw.data)
genes <- rownames(rawdata)
```

## Identify cell doublets using scrublet
Identify cell doublets using scrublet in python. The python script calculates doublet scores and produced UMAP plots of cells colored by doublet score. The doublet threshold is set to a minimum of either the automatic threshold or 0.3. I decided this empirically, but it is something that you may want to change. Remember, the doublet scores can always be used later to change what is called as a doublet.
```{r scrublet, eval = T}
source_python(paste0(proj.path, "/src/scrublet_function_GTEx.py"))
for (i in seq(sampleid.cumulus)) {
  run_scrublet(sampleid.cumulus[i], sparse.dir[i], figures.pythondir)
}
```

## Read scrublet doublet data into Seurat object
This code reads in the doublet scores produced by scrublet and also allows us to remove doublets from the Seurat object, although I do not advise doing this until you look where the doublets occur among the clustered cell subsets. The two lines to remove doublets are at the end of this code chunk and they can be commented out if you want to keep the called doublets as cells in the analysis.
```{r read_scrublet, eval = T}
# load(Rda.path)  # load in case script is beginning from here
for (i in seq(sampleid.cumulus)) {
  # Read scrublet file with cell doublet assignments and doublet scores.
  scrubletfile <- paste0(figures.dir, sampleid.cumulus[i], "_scrublet_scores.txt")
  scrublet <- read.table(scrubletfile, sep = "\t", header = T, row.names = NULL)
  scrublet$doublet <- as.logical(scrublet$doublet)

  gcdata[[i]]@meta.data[["doublet"]] <- scrublet$doublet
  gcdata[[i]]@meta.data[["doubletscore"]] <- scrublet$doubletscore
  p <- VlnPlot(gcdata[[i]], c("nUMI", "nGene"), group.by = "doublet", point.size.use = 0.1,
               do.return = T)
  save_plot(paste0(figures.path[i], 'nUMI_nGene_scrubletfilter.png'), p, ncol = 2,
            base_aspect_ratio = 0.7, base_height = 10)
  print(p)

  n.doublet <- sum(scrublet$doublet)
  p.doublet <- round(100*n.doublet/length(scrublet$doublet), 2)
  print(paste0("% doublet = ", p.doublet, ", # doublets = ", n.doublet))

  # Remove doublets from Seurat object. Comment this out if you do not want to remove doublets.
  # cells.singlet <- rownames(gcdata[[i]]@meta.data[!gcdata[[i]]@meta.data$doublet, ])
  # gcdata[[i]] <- SubsetData(gcdata[[i]], cells.use = cells.singlet)
}

# Save current progress with filtering cells and doublets from each sample.
save(gcdata, sampleid.cumulus, file = Rda.cumulus.scrublet.path)
```

## Plot doublets and empty drops on tSNE/UMAP 
```{r plot doublets and empty drops, eval = T}
# Add empty drops information to seurat objects
load(Rda.emptydrops.path)
load(Rda.cumulus.scrublet.path)

# Add the sample name to the empty drops cell names
cells.keep.emptydrops.longname <- lapply(names(cells.keep.emptydrops), function(x) paste0(x, '-', cells.keep.emptydrops[[x]]))
names(cells.keep.emptydrops.longname) <- sampleid.counts

# combine into one long list - how do I do this without a loop?
cells.keep.emptydrops.longname.all <- c()
for (i in sampleid.counts){
        cells.keep.emptydrops.longname.all <- c(cells.keep.emptydrops.longname.all, cells.keep.emptydrops.longname[[i]])
}

# Make data frame of empty drops information for adding to metadata
for (id in sampleid.cumulus){
        df.ed <- data.frame(emptydrop=rep(TRUE, nrow(gcdata[[id]]@meta.data)))
        rownames(df.ed) <- rownames(gcdata[[id]]@meta.data)
        df.ed[which(rownames(df.ed) %in% cells.keep.emptydrops.longname.all),1] <- FALSE
        gcdata[[id]] <- AddMetaData(gcdata[[id]], df.ed)
}


# Plot empty drops and doublets
for (id in seq(sampleid.cumulus)){
        # empty drops
        df <- data.frame(gcdata[[id]]@dr$tsne@cell.embeddings, emptydrop = gcdata[[id]]@meta.data$emptydrop)
        p1 <- ggplot(data = df) + geom_point(mapping = aes(tSNE_1, tSNE_2, color = emptydrop), size = 0.5) +
          labs(color = "Emptydrop identity") + scale_colour_manual(values = material.heat(2)) +
          guides(colour = guide_legend(override.aes = list(size=5))) +
                theme(aspect.ratio = 1, text=element_text(size=8),
                      axis.text = element_text(size=8))
        ggsave(paste0(figures.path[id], "emptydrops_tSNE.pdf"), plot=p1, device = "pdf", width = 4, height = 4, dpi = 200)

        # doublets
        df <- data.frame(gcdata[[id]]@dr$tsne@cell.embeddings, doublet = gcdata[[id]]@meta.data$doublet)
        p1 <- ggplot(data = df) + geom_point(mapping = aes(tSNE_1, tSNE_2, color = doublet), size = 0.5) +
          labs(color = "Doublet identity") + scale_colour_manual(values = material.heat(2)) +
          guides(colour = guide_legend(override.aes = list(size=5))) +
                theme(aspect.ratio = 1, text=element_text(size=8),
                      axis.text = element_text(size=8))
        ggsave(paste0(figures.path[id], "doublet_tSNE.pdf"), plot=p1, device = "pdf", width = 4, height = 4, dpi = 200)
}

# save seurat object with empty drops and doublet iinfo in metadata
save(gcdata, sampleid.cumulus, file = Rda.cumulus.scrublet.emptydrops.path)


```


## Use SingleR to identify cell types 
```{r SingleR, eval = T}
library(SingleR)
library(data.table)

get_annotations=function(singler,suffix=""){
        anno=as.data.table(singler[c("cell.names","labels","scores","pval")])
        setnames(anno,"labels.V1","labels")
        anno[,score:=get(paste0("scores.",labels)),by=1:nrow(anno)]
        anno[,rel_score:=score/rowSums(.SD), .SDcols = grep("scores",names(anno))]
        anno[,N_labels:=.N,by=labels]
        anno_df=data.frame(anno[,-c(grep("scores.",names(anno),value = TRUE),"cell.names"),with=FALSE],row.names = anno$cell.names)
        colnames(anno_df)=paste0(colnames(anno_df),suffix)
        return(anno_df)
}

for (i in seq(sampleid.cumulus)) {

  ## Create SingleR object from Seurat object
  newsingler <- CreateSinglerObject(gcdata[[i]]@raw.data, annot = gcdata[[i]]@meta.data$louvain, gcdata[[i]]@project.name, min.genes = 0, 
                                    technology = "10X", species = "Human", ref.list = list(), normalize.gene.length = F,
  variable.genes = "de", fine.tune = F, do.signatures = F, clusters = gcdata[[i]]@meta.data$louvain, do.main.types = T, reduce.file.size = T, numCores = 15)
  
          newsingler$seurat = gcdata[[i]]
  
          labels_detail = get_annotations(newsingler$singler[[1]]$SingleR.single,"_detail")
          labels = get_annotations(newsingler$singler[[1]]$SingleR.single.main,"")
          newsingler$seurat = AddMetaData(newsingler$seurat,labels_detail)
          newsingler$seurat = AddMetaData(newsingler$seurat,labels)
  
          sub.data <- SubsetData(newsingler$seurat, subset.name="N_labels",accept.low=30)
  
          # Color tSNE plot by annotation
          df <- data.frame(sub.data@dr$tsne@cell.embeddings, nUMI = sub.data@meta.data$nUMI,
                           nGene = sub.data@meta.data$nGene, ident = sub.data@meta.data$labels)
          n.clusters <- length(unique(df$ident))
          centers <- df %>% group_by(ident) %>% summarise(x = median(tSNE_1), y = median(tSNE_2))  # cluster labels
          p1 <- ggplot(data = df) + geom_point(mapping = aes(tSNE_1, tSNE_2, color = ident), size = 0.5) +
                  labs(color = "cluster") + #+ scale_colour_manual(values = material.heat(n.clusters))
            geom_text(data = centers, mapping = aes(x = x, y = y, label = ident), size = 4) +
          guides(colour = guide_legend(override.aes = list(size=5))) +
                  theme(legend.text = element_text(size=8), legend.title = element_blank()) +
                  theme(axis.text.x = element_text(size=8),
                        axis.text.y = element_text(size=8),
                        axis.title.x = element_text(size=8),
                        axis.title.y = element_text(size=8))
  
  
          ggsave(paste0(figures.path[i], "SingleR_tSNE_thresh30.png"), width = 10, height = 6)
}


```